### ISSUE TITLE
[css-properties-values-api] is initialValue required when syntax is not "*"?

### ISSUE THREAD

**Issue by heycam** [w3c/css-houdini-drafts#286]
I think the `initialValue` field on a PropertyDescriptor is effectively required when the syntax string is `"*"`, but the spec doesn't say what to do when it's not specified.  An exception should probably be thrown.

**Comment by tabatkins** [comment]
Nah, it's fine to be without - it's just like an unregistered custom property then, defaulting to an invalid empty value.  This is now specified.

**Comment by jyc** [comment]
It looks like this was changed (though not mentioned in the commit message) in commit 291eb20f38fbb4142ad4a1aba3bdb485ed93f78a. Was this intentional?

**Comment by tabatkins** [comment]
How was it changed? Step 3 handles the case where the syntax a "*"/omitted and the initial value is omitted.

**Comment by jyc** [comment]
Before: If `syntax` is not "*" and `initialValue` is not specified, it defaults to an invalid empty value, per your comment in https://github.com/w3c/css-houdini-drafts/issues/286#issuecomment-246861212. Because you were disagreeing with heycam, I assumed you implied that an exception should not be thrown, and that this value should just trigger the "invalid at computed-value time" behavior when substituting, and be return when we get it through OM.

After: If `syntax` is not "*" and `initialValue` is not specified, then syntax becomes `parsed syntax`. We do not go to the following case because `parsed syntax` is not undefined (i.e. `syntax` was "*" in step 3):

> If parsed syntax is undefined, and initialValue is not present, let parsed initial value be empty. This must be treated identically to the "default" initial value of custom properties, as defined in [css-variables]. Skip to the next step of this algorithm.

We do not go to the following case either, again because `parsed syntax` is not undefined:

> Otherwise, if parsed syntax is undefined, parse initialValue as a <declaration-value>. If this fails, throw a SyntaxError and exit this algorithm. Otherwise, let parsed initial value be the parsed result. Skip to the next step of this algorithm.

Instead, we go to this case:

> Otherwise, if initialValue is not present, throw a SyntaxError and exit this algorithm.

Which contradicts what you were saying about not throwing an exception.

**Comment by tabatkins** [comment]
In step 2:

> If |syntax| is not present, or is equal to <code>`"*"`</code> (U+002A ASTERISK), let |parsed syntax| be undefined, and skip to the next step of this algorithm.

Thus either omitting the syntax or explicitly specifying `"*"` makes |parsed syntax| be undefined, hitting the appropriate clause in step 3 that allows an omitted initial value.

**Comment by jyc** [comment]
The issue is about when syntax is not "*", not when the syntax is not omitted or when the syntax is explicitly specified as `"*"`.

**Comment by jyc** [comment]
Oops, actually, I think I've mistaken this issue with another one. Checking now.

**Comment by tabatkins** [comment]
The issue's title is a little misleading; the issue that heycam raised is explicitly about an unspecified `syntax`, not a specified `syntax` with a value other than `"*"`.

**Comment by jyc** [comment]
Yep, you are correct. Sorry for the misunderstanding, thanks for the help!

To confirm, then, when `syntax` is either omitted or `"*"`, the intent is that `initialValue` is required? I had been imagining it would default to the same initial value as unregistered custom properties, and would then end up invalid in a `var()` reference if nothing was specified or inherited. Seems like that would make it a little simpler for people who are just using a registered custom property for a few specific elements.

**Comment by tabatkins** [comment]
No, the exact opposite!  When it's omitted or `"*"`, then step 3 allows initialValue to be unspecified without raising an error; in that case, the initial value is set to an initially invalid value, just like any unregistered custom property.

If the spec is in any way unclear on this point, let me know, but as far as I can tell this behavior is completely specified.

**Comment by jyc** [comment]
My bad, still tired from traveling yesterday, that is what I meant to say (edited the sentence but forgot to negate). Thanks! The spec is very clear, thanks for the work on it.

I still am wondering about the second part of that question, though. It does seem to me like it'd be really useful to allow omitting an `initialValue` with the same behavior behavior for properties whose syntax is specified as something other than `"*"` as well. Personally, it was something I expected & used while writing my testcases and demos. Has this already been discussed, or could I make another issue asking about a spec change?

**Comment by tabatkins** [comment]
That doesn't work, because then you're no longer guaranteed that the value you get when you ask for the property will match the specified syntax.  Like, a property specified to be a `<color>` might not return a color when you asked for the computed value.

It's also *most likely* an authoring error, just accidentally leaving off the initial value.  If we do think allowing it is important, we can fold it into expansions to allowed syntax in the future.

**Comment by jyc** [comment]
If omitting the initial value were allowed, then it seems like it'd make sense to expect that getting a property specified to be a `<color>` through OM or Typed OM would return `""` or `null` or some sort of other invalid value–to me that would seem more natural than every element having some property that I only wanted to use to specify how some specific sort of element is painted or  animated. As the spec is right now it does seem kind of weird, though. Also, I agree that it could make people confused. Sounds fair to me to make it a potential expansion for the future, though, and thanks for the help. Is there somewhere I should write this to record that?

**Comment by tabatkins** [comment]
If you feel like it's something to be addressed, just open another issue for that. ^_^

**Comment by jyc** [comment]
Cool, thanks again for the help! Really appreciate it.
